### PR TITLE
Shows snackbar when no schedules can be generated

### LINKED ISSUE #278: Add snackbar when no schedules are generated to frontend
Per Adel's comment [here](https://github.com/aggie-coding-club/Rev-Registration/pull/255#pullrequestreview-430249486), we should add a snackbar popup whenever no schedules are generated (when `/scheduler/generate` returns an empty array)

### PR DESCRIPTION
**Pull request by firejake308**
## Description

I often hit "Generate Schedules" and then sit for a very confusing minute wondering why it's taking so long, until I realize that I didn't check "Include full sections". Users will almost certainly make similar mistakes, so we need an error handling method to let them know that they need to change their filters. To this end, I have added a snackbar that shows when the Generate Schedules  button returns no schedules.

***tests coming soon***

## Rationale

I think the only possible controversial thing here would be the text in the error bar. It currently reads "No schedules found. Try widening your criteria." Let me know if you'd like something else.

## How to test

Hitting Generate Schedules without adding any courses is the easiest way to test this.

## Screenshots

image

## Related tasks

Closes #278

### CODE REVIEW

File: autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx [hunk 1]
```diff
@@ -117,6 +129,17 @@ const ConfigureCard: React.FC = () => {
             : 'Generate Schedules'}
         </Button>
       </div>
+      <Snackbar
+        open={showSnackbar}
+        autoHideDuration={10000}
+        message="No schedules found. Try widening your criteria."
+        onClose={(): void => setShowSnackbar(false)}
```
Comment (ryan-conn) [inline autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:136]: Might want to change this callback to the following:
```ts
  const handleSnackbarClose = (_event: any, reason: string): void => {
    if (reason === 'clickaway') return;
    setShowSnackbar(false);
  };
```
Right now the snackbar closes if you click anywhere outside of it, which makes it easy to ignore if you're not paying close attention

File: autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx [hunk 2]
```diff
@@ -117,6 +129,17 @@ const ConfigureCard: React.FC = () => {
             : 'Generate Schedules'}
         </Button>
       </div>
+      <Snackbar
+        open={showSnackbar}
+        autoHideDuration={10000}
```
Comment (ryan-conn) [inline autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:134]: Like Gannon said in the meeting, we'll probably want to change this to ~5-6 seconds

File: autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx (hunk 1 above)
Comment (firejake308) [reply autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:136]: Hmm, I know this was the recommended behavior in the docs, but I kinda liked having it go away if you clicked outside. Can I get a 3rd opinion from @gannonprudhomme or @ThatJuanGuy ?

Comment (ThatJuanGuy) [reply autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:136]: I pulled the branch but I can't really check how obvious the snack bar is when it pops up because I still generate the same 3 preset schedules regardless of what courses I have selected. Regardless, I don't think it should disappear if you click outside of it. This is because if no schedules are showing up the user might be randomly clicking to pass the time or repeatedly clicking the generate button to try and make it go faster. Both of which would make it much easier for them to miss the message letting them know no schedules can be generated with their current settings.

File: autoscheduler/frontend/src/tests/ui/ConfigureCard.test.tsx [hunk 3]
```diff
@@ -153,4 +153,47 @@ describe('ConfigureCard component', () => {
       expect(sections.length).toEqual(0);
     });
   });
+
+  describe('shows an error message', () => {
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/tests/ui/ConfigureCard.test.tsx:157]: nit: This should probably be `shows an error snackbar`. Same with the `does not show an error message`.

File: autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx [hunk 4]
```diff
@@ -89,7 +107,7 @@ const ConfigureCard: React.FC = () => {
         <div id={styles.cardHeader}>Configure</div>
       }
     >
-      <div className={styles.buttonContainer}>
+      <div className={styles.buttonContainer} ref={ref}>
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:110]: Just for my understanding, what does adding this ref do for us?

Comment (firejake308) [reply autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:110]: I'll add a comment above the initial `useRef` because you're right that it's not immediately clear what its purpose is. Basically, when I was running tests, I kept getting an error about React dispatching a `setState` in an async callback after un-mounting a component. My solution to this was to obtain a reference to the DOM element and check if it's still mounted before calling any asynchronous `setState()`s.

TL;DR the ref fixes async setState
